### ISSUE TITLE
Option to use a global sender address for emails (to fight spamfilters?)

### ISSUE THREAD

**Issue by ajnyga** [pkp/pkp-lib#4014]
Currently OJS and OMP will send out emails so that the sender address is the actual address of for example the editor making a decision. 

In some cases this could lead to a situation where the email is marked as spam. Especially, if the sender and the receiver are from the same institution and use the same email server (which is probably quite common in many cases).

Suggestion:
- additional global_sender setting in config.inc.php
- if the setting is filled in with an email address, OJS will use that address for sending all emails. OJS installations can the create an actual email address (like no-reply@myinstitution.com) on their own email server and just make sure that the server keeps whitelisted
- the user email in OJS is used only in the reply-to field and the name of the sender
- additionally the sender name could be changed to something like "John Doe via sitename.com", because some users may be suspicious if they see a sender name mixed with an uknown email address

This is probably something that only requires small changes around here https://github.com/pkp/pkp-lib/blob/master/classes/mail/Mail.inc.php#L306 where the code checks for the setting in config.inc.php and if found, replaces the email address and sets the reply-to field.

Thoughts?

**Comment by forgive38** [comment]
Hi all,

We need exactly this functionality
Thank you

**Comment by asmecher** [comment]
@ctgraham, you've been fighting with DMARC etc. for quite some time and I think you're the best authority we've got on the subject -- do you think this would be a good long-term approach?

My sense is that it would, and might also facilitate some things we'd like to approach -- e.g. with a dedicated email account for the journal, we can start experimenting with getting OJS to consume email and not just send it. (Think responding directly to emails on discussions, for example.)

**Comment by asmecher** [comment]
(FYI @mfelczak)

**Comment by ctgraham** [comment]
@ajynga, the `default_envelope_sender` and `force_default_envelope_sender` in config.inc.php should already function this way, [courtesy of `Mail::getEnvelopeSender()`](https://github.com/pkp/pkp-lib/blob/22837fc09f5f95c21434095d842925d61612879e/classes/mail/Mail.inc.php#L489) and the [configuration check there](https://github.com/pkp/pkp-lib/blob/master/classes/mail/Mail.inc.php#L84-L86).

The `Reply-To` component is missing, but would readily be added [in the `Mail::Send` method](https://github.com/pkp/pkp-lib/blob/22837fc09f5f95c21434095d842925d61612879e/classes/mail/Mail.inc.php#L489).

If user addresses cannot be trusted implicitly, the best near-term and mid-term configuration is the use of:
```
Sender: DMARC trusted address
From: user address
Reply-To: user address
```

The best log-term configuration would use an email reply processor (like GitHub's, for example) where the reply to was a handler address + a hash representing the conversation.  I don't know of a open source tool which provides this functionality via API currently.

I'm actively working on testing this with OJS 2.x and will plan on reconciling the final changes with 3.x.

**Comment by ajnyga** [comment]
hah, yeah noticed that as I was testing this. 
I think the reply-to part would be important.

**Comment by ajnyga** [comment]
This is the quick code I wrote:
```
	/**
	 * Set the sender of the message.
	 * @param $email string
	 * @param $name string optional
	 */
	function setFrom($email, $name = '') {
		if (Config::getVar('email', 'global_sender')) {
			$globalSender = Config::getVar('email', 'global_sender');
			$globalSenderName = $name . " via " . $_SERVER['HTTP_HOST']; 
			$this->setData('from', array('name' => $globalSenderName, 'email' => $globalSender));
			$this->setReplyTo($email, $name);
		}
		else {
			$this->setData('from', array('name' => $name, 'email' => $email));
		}
	}
```

**Comment by ctgraham** [comment]
Just a note: This shouldn't live in the setter method.  The programmer should be able to construct the `Mail` object normally without the side effects of setting conditional properties based on the larger configuration. Then, when the sending action is effected, the configuration treats the `Mail` object in special ways to effectively route the message.

**Comment by ajnyga** [comment]
sure, that was just the quickest way of testin this.

**Comment by ajnyga** [comment]
Ok, so the `force_default_envelope_sender` setting works really nice. I added my own email there on one of our test servers and sent an email to gmail from OJS. Authentication results in the message header are:

```
ARC-Authentication-Results: i=1; mx.google.com;
       spf=pass (google.com: best guess record for domain of my@myorganisation.fi designates 8x.2xx.2xx.xx as permitted sender) smtp.mailfrom=my@myorganisation.fi;
       dmarc=fail (p=NONE sp=NONE dis=NONE) header.from=emailhere.fi
Return-Path: <my@myorganisation.fi>
Received: from promail.myorganisation.fi (mail.myorganisation.fi. [8x.2xx.2xx.1x])
        by mx.google.com with ESMTPS id xxx-xxxxx8x2x4xg.xx.2xx8.0x.27.1x.x.x7
        for <receiver_address_here@gmail.com>
        (version=TLS1_2 cipher=ECDHE-RSA-AES128-GCM-SHA256 bits=128/128);
        Mon, 27 Aug 2018 11:14:27 -0700 (PDT)
Received-SPF: pass (google.com: best guess record for domain of my@myorganisation.fi designates xx.2xx.2xx.xx as permitted sender) client-ip=xx.xx9.xx2.xx;
Authentication-Results: mx.google.com;
       spf=pass (google.com: best guess record for domain of my@myorganisation.fi designates xx.2xx.xxx.xx as permitted sender) smtp.mailfrom=my@myorganisation.fi;
       dmarc=fail (p=NONE sp=NONE dis=NONE) header.from=myorganisation.fi
```
Basically SPF was failing before I added the envelope_sender, but DMARC is still failing. Any idea why @ctgraham? And maybe we should add this whole thing a a FAQ to the forum and close this ticket?

**Comment by mfelczak** [comment]
"The best long-term configuration would use an email reply processor (like GitHub's, for example) where the reply to was a handler address + a hash representing the conversation. I don't know of a open source tool which provides this functionality via API currently."

Agreed with @ctgraham

When I look at other 3rd-party services that I use, I see the following:

From: FirstName LastName (via service) <activity+id@domain.com>
Reply-to: <activity+id-reply@domain.com>

That should allow the journal to setup SPF and/or DKIM for their domain to meet DMARC requirements and would do away with the current misalignment between the FROM and ENVELOPE FROM domains.

**Comment by ctgraham** [comment]
@ajnyga, you haven't included a DKIM header in your sample output.  Are you using DKIM signing at the server level?

If both DKIM and SPF are in place, the next step is to review the DMARC statement for the domain, probably set to just reporting. 

One reporting has been tested, you can upgrade the DMARC statement to reject or quarantine messages.

https://dmarc.org/overview/

**Comment by ajnyga** [comment]
I have to discuss this with the guy handling our servers, I do not have access to the email server (or know these kinds of details to begin with). But thank @ctgraham, I will ask about these and see if we can get those fixed.

I could close this, but maybe you should start a new issue on the github like email posting functionality?

**Comment by asmecher** [comment]
I'd suggest keeping this one open -- the discussion here is useful.

**Comment by ajnyga** [comment]
Sure, feel free to revise the title and the first message there if needed. 

When I have our own server working the way I would like, I could do a FAQ about using the force_default_envelope_sender setting to the forum.

So we could dedicate this issue for the github-like functionality @ctgraham mentioned.

**Comment by mfelczak** [comment]
@ajnyga, did you have any luck with DMARC with your tweaks to setFrom?

**Comment by asmecher** [comment]
@ajnyga, I'd love your comment on the changes from @ctgraham that I just merged into the `master` branch! See https://github.com/ulsdevteam/pkp-lib/commit/625a86e0f3f6b50c62bc202140dcd1fa20cdee1c for details.

**Comment by ajnyga** [comment]
Should the reply-to field in the received message now contain both the email of the sender and the email defined in the envelope_sender setting? 

I tried to send a message from OJS first by using the original code (3.1.1.2) and the changed code here. The receiving end was an gmail account, the sending user had a finnish university account and the envelope sender was a real email account on the email server OJS is also using. 

I then checked the details of both these messages from the gmail account and all the sender related information is identical. Should there be a visible difference there somewhere? I am maybe missing something here?

**Comment by ctgraham** [comment]
The intent is for the From address to be added as a Reply-To, if it differs from the Envelope Sender.  The envelope sender may be sent as the "bounce address" within a journal, or may be set from config.inc.php.

Let's say that OJS is configured with:
```
allow_envelope_sender = On
default_envelope_sender = noreply@mydomain.tld
force_default_envelope_sender = On
```

And that the following mail is created:
```
From: editor@gmail.com
To: author@yahoo.com
Reply-to: sectioneditor@an.edu
```

The resulting SMTP headers might look like:
```
Return-path: noreply@mydomain.tld
Sender: noreply@mydomain.tld
From: editor@gmail.com
To: author@yahoo.com
Reply-to: sectioneditor@an.edu, editor@gmail.com
Authentication-Results: spf=pass ...
```
Setting the `Return-Path` and preserving the `Sender` is the option/responsibility of the receiving MTA and not guaranteed.  The `Authentication-Results` SPF analysis should reflect the authentication of `noreply@mydomain.tld`.

**Comment by ajnyga** [comment]
@asmecher what if we change the name of the issue to: "Introduce an email reply processor and enable users to participate to discussions by answering emails"

Because I think that my original issue is solved here. Or we could just close this and create a new one with Clinton's suggestion. I think that having a feature like that would move OJS to whole new level.

**Comment by asmecher** [comment]
@ajnyga, I wholeheartedly support that idea -- but I do think it's a bigger discussion than this issue. I've filed it here: https://github.com/pkp/pkp-lib/issues/4142

@ctgraham and @ajnyga, are the intentions of this original issue addressed by the force-envelope-sender option and the tweaks Clinton has made?

**Comment by ajnyga** [comment]
I think they are. The rest is up to our own email server settings.

**Comment by asmecher** [comment]
Thanks, @ajnyga! Closing, unless @ctgraham thinks otherwise.

**Comment by ajnyga** [comment]
Just fyi that I created a plugin that replaces all the from fields and uses the envelope_sender field as the sender. https://github.com/ajnyga/useGlobalSender

It supports the master branch version of mail.inc.php (excluding this pr from @asmecher https://github.com/pkp/pkp-lib/commit/38a6c7a9e35278a9e7c28957fc5d61ddcbca3122 because it seemed to cause an error in our installations)

**Comment by asmecher** [comment]
> excluding this pr from @asmecher 38a6c7a because it seemed to cause an error in our installations

Could you try updating your Composer dependencies and see if that resolves the error you were having with 38a6c7a?

**Comment by ajnyga** [comment]
sorry, can not do that in our production because we do not use git upgrades there. But that is probably the reason.
